### PR TITLE
CMake: Avoid temporary build outputs outside of build directory

### PR DESCRIPTION
**Pull request by maleadt**
... by changing the working directory, or specifying `-o`.

Also includes a couple of fixes wrt. exit code checking, as encountered when debugging build failures on Windows.

### CODE REVIEW

File: cmake/LLVM.cmake [hunk 1]
```diff
@@ -397,23 +401,37 @@ endmacro()
...
+  file(REMOVE "${RANDOM_FILENAME}")
 endmacro()
 
 # clang++ try-link macro
 macro(custom_try_link_clang SOURCE1 SOURCE2 RES_VAR)
-  set(RANDOM_FILENAME "${CMAKE_BINARY_DIR}/compile_test_${RNDNAME}.${SUFFIX}")
-  custom_try_compile_c_cxx_silent("${CLANG}" "c" "${SOURCE1}" "${SOURCE2}" ${RES_VAR}  "-o" "${RANDOM_FILENAME}" ${ARGN})
+  string(RANDOM RNDNAME)
+  if(WIN32)
+    set(RANDOM_FILENAME "${CMAKE_BINARY_DIR}/compile_test_${RNDNAME}.exe")
```
Comment (linehill) [inline cmake/LLVM.cmake:430]: Why [${SUFFIX}](https://cmake.org/cmake/help/v3.0/prop_tgt/SUFFIX.html) does not work here? Perhaps the problem here is that it should be used as `... ${RNDNAME}${SUFFIX}` (no dot).

File: cmake/LLVM.cmake [hunk 2]
```diff
@@ -432,7 +450,7 @@ macro(custom_try_run_exe SOURCE1 SOURCE2 OUTPUT_VAR RES_VAR)
     execute_process(COMMAND "${OUTF}" RESULT_VARIABLE RESV OUTPUT_VARIABLE ${OUTPUT_VAR} ERROR_VARIABLE EV)
     set(${RES_VAR} ${RESV})
     file(REMOVE "${OUTF}")
-    if(${RESV})
+    if(NOT ${RESV} EQUAL 0)
```
Comment (linehill) [inline cmake/LLVM.cmake:453]: What is wrong with `if(${RESV})` here?

Comment (maleadt) [reply cmake/LLVM.cmake:453]: On Windows/MinGW, I got string values in `RESV` (e.g. `Access violation`, not simple exit codes), making this not behave as intended.

File: cmake/LLVM.cmake (hunk 1 above)
Comment (maleadt) [reply cmake/LLVM.cmake:430]: `SUFFIX` is not an argument to this macro, which it is with other macros (like `custom_try_compile_c_cxx_silent` or `custom_try_compile_c_cxx`). I don't see how the linked target property is relevant here?

Comment (linehill) [reply cmake/LLVM.cmake:430]: I thought the SUFFIX was a predefined CMake variable and didn't read the description carefully.

Probably the SUFFIX was mixed up with CMAKE_EXECUTABLE_SUFFIX. Could you use the latter?

**Review by linehill — APPROVED** [review]

**Pull request merged**